libraries/chain/controller.cpp

```diff
@@ -987,6 +994,10 @@ struct controller_impl {
    vote_signal_t                             voted_block;     // emitted when a local finalizer votes on a block
    vote_signal_t                             aggregated_vote; // emitted when a vote received from the network is aggregated
 
+   std::function<void(produced_block_metrics)> _update_produced_block_metrics;
+   std::function<void(speculative_block_metrics)> _update_speculative_block_metrics;
+   std::function<void(incoming_block_metrics)> _update_incoming_block_metrics;
```

Comment:
Any reason why these are not signals?

Reply:
They were moved from `producer_plugin`. I think the main reason they are not signaled is you don't want to populate them if prometheus is not enabled. Signals are aways signaled (as far as I know), so would need to create the struct to signal it. Or would need some other mechanism to know if we should signal or not.

Reply:
You could use [signal::empty()](https://www.boost.org/doc/libs/1_86_0/doc/html/boost/signals2/signal.html)  which is true if no slots are connected to the signal.

Reply:
I was looking for that, didn't look in the right place apparently. That would require a mutex I assume. Regardless this is all going away rather soon, so would prefer to just leave as is.

Reply:
That's fine, you can ignore my comments related to these metrics callbacks.